### PR TITLE
[1.2.0] SQL Docs

### PR DESCRIPTION
**Pull request by yuce**
Just adds documentation for the SQL bits.

### CODE REVIEW

File: sql/driver/doc.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+	cfg := hazelcast.Config{}
+	cfg.Cluster.Name = "prod"
+	cfg.Serialization.SetPortableFactories(&MyPortableFactory{})
+	db := driver.Open(cfg)
+
+Executing Queries
+
+database/sql package supports two kinds of queries: The ones returning rows (select statements and a few others) and the rest (insert, update, etc.).
+The former kinds of queries are executed with QueryXXX methods and the latter ones are executed with ExecXXX methods of the sql.DB instance returned from sql.Open or driver.Open
```
Comment (srknzl) [inline sql/driver/doc.go:106]:
```suggestion
The former kinds of queries are executed with QueryXXX methods and the latter ones are executed with ExecXXX methods of the sql.DB instance returned from sql.Open or driver.Open.
```

File: sql/driver/doc.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	var name string
+	var age int
+	for rows.Next() {
+		err := rows.Scan(&name, &age)
+		// handle the error
+		fmt.Println(name, age)
+	}
+
+Context variants of Query and Exec, such as QueryContext and ExecContext are fully supported.
+Context variants are used to pass Hazelcast specific parameters.
```
Comment (srknzl) [inline sql/driver/doc.go:136]: What does this mean? I did not get it. 

EDIT: I understood this after reading Passing Hazelcast Specific Parameters section. Maybe we can add a reference(or link) to that section here.

File: sql/driver/example_test.go [hunk 3]
```diff
@@ -25,7 +25,7 @@ import (
 )
 
 func Example() {
```
Comment (srknzl) [inline sql/driver/example_test.go:27]: Is this function used 🤔 anywhere?  Also why add random token and cluster name?

File: sql/driver/doc.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+	// set the cursor buffer size to 10_000
+	ctx := driver.WithCursorBufferSize(context.Background(), 10_000)
+	// set the query timeout to 2 minutes
+	ctx = driver.WithQueryTimeout(ctx, 2*time.Minute)
+	// use the parameters above with any methods that uses that context
+	rows, err := db.QueryContext(ctx, "select * from people")
+
+Creating a Mapping
+
+To connect to a data source and query it as if it is a table, a mappings should be created.
```
Comment (srknzl) [inline sql/driver/doc.go:158]:
```suggestion
To connect to a data source and query it as if it is a table, a mapping should be created.
```

File: sql/driver/doc.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+Currently, mappings for Map, Kafka and file data sources are supported.
+
+You can read the details about mappings here: https://docs.hazelcast.com/hazelcast/latest/sql/sql-overview#mappings
+
+Supported Data Types
+
+The following data types are supported, as both builtin values and in types supporting the portable interface when inserting/updating:
```
Comment (srknzl) [inline sql/driver/doc.go:165]: I think this part can be clarified. I think you mean you list builtin types and their corresponding names in portable. Actually real, decimal, date etc. are not portable names, they are hazelcast's SQL data types, aren't they?

File: sql/driver/doc.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+	- string (varchar)
+	- int8 (tinyint)
+	- int16 (smallint)
+	- int32 (integer)
+	- int64 (bigint)
+	- bool (boolean)
+	- float32 (real)
+	- float64 (double)
+	- types.Decimal (decimal)
+	- time.Time (date) Detected by checking: hour == minute == second == nanoseconds = 0
```
Comment (srknzl) [inline sql/driver/doc.go:176]: where is time? also the constraints seem wrong. For example in timestamp with time zone `year` should not be limited to 0, likewise for timestamp.

File: sql/driver/doc.go [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+        CREATE MAPPING person (
+			__key BIGINT,
+			age BIGINT,
+			name VARCHAR
+		)
+        TYPE IMAP
+        OPTIONS (
+            'keyFormat' = 'bigint',
+            'valueFormat' = 'json-flat'
+        )
```
Comment (srknzl) [inline sql/driver/doc.go:225]: I see this code sample as indented, I don't know why. 
image


all other code samples are rendered normally.

File: sql/driver/doc.go [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+	- log: One of the following: off, fatal, error, warn, info, debug, trace. Default: info. Example: log=debug
+	- cluster.name: A string. Specifies the cluster name.Default: dev. Example: cluster.name=hzc1
+	- cloud.token: A string. Sets the Hazelcast Cloud token. Example: cloud.token=1234567abcde
+	- stats.period: Duration, which can be parsed by time.Parse.
+	  Use one of the following suffixes: s (seconds), m (minutes), h (hours). Example: 10s
+	- ssl: A boolean. Enables/disables SSL connections. Defaults: false. Example: ssl=true
+	- ssl.ca.path: The path to the PEM file for the certificate authority. Implies ssl=true. Example: ssl.ca.path=/etc/ssl/ca.pem
+	- ssl.cert.path: The path to the TLS certificate. Implies ssl=true. Example: ssl.cert.path=/etc/ssl/cert.pem
+	- ssl.key.path: The path to the certificate key. Implies ssl=true. Example: ssl.key.path=/etc/ssl/key.pem
+	- ssl.key.password: The optional certificate password. Example: ssl.key.password=m0us3
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/doc.go:68]: That's right, if ssl.cert.path is given, ssl is set to true. SSL key password is optional, and it is not very useful by itself.

File: sql/driver/doc.go (hunk 2 above)
Comment (yuce) [reply sql/driver/doc.go:136]: Cross-references are not supported in Go, but I will try to explain in that better.

File: sql/driver/example_test.go (hunk 3 above)
Comment (yuce) [reply sql/driver/example_test.go:27]: That example appears in the godoc. They are just placeholders.

File: sql/driver/doc.go (hunk 5 above)
Comment (yuce) [reply sql/driver/doc.go:165]: That's right. I'll try to clarify that.

File: sql/driver/doc.go (hunk 6 above)
Comment (yuce) [reply sql/driver/doc.go:176]: That's right, copy&paste error.

File: sql/driver/doc.go (hunk 7 above)
Comment (yuce) [reply sql/driver/doc.go:225]: My bad, extra indentation.

File: sql/driver/doc.go [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+
+	db, err := sql.Open("hazelcast", "hz://@localhost:5701?cluster.name=dev")
+
+Use hazelcast as the driver name.
+The DSN may be blank. In that case, the default configuration is used.
+Otherwise, the DSN must start with the scheme (hz://) and have the following optional parts:
+
+	- Username and password for the cluster, separated by a column: dave:s3cr3t
+	- Hazelcast member addresses, separated by commas: server1:port1,server2:port2
+	- Options as key=value pairs, separated by ampersond (&). Both the key and value must be URL encoded: cluster.name=dev&ssl=true
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+Package driver provides a standard database/sql compatible SQL driver for Hazelcast.
+
+This driver supports Hazelcast 5.0 and up. Check out the Hazelcast SQL documentation here: https://docs.hazelcast.com/hazelcast/5.0/sql/sql-overview
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+There should be a question mark (?) between the address(es) and options.
+Here is a full DSN:
+
+	hz://dave:s3cr3t@my-server1.company.com:5000,my-server2.company.com:6000?cluster.name=prod&ssl=true&log=warn
+
+The following are the available options:
+
+	- unisocket: A boolean. Enables/disables the unisocket mode. Default: false. Example: unisocket=true
+	- log: One of the following: off, fatal, error, warn, info, debug, trace. Default: info. Example: log=debug
+	- cluster.name: A string. Specifies the cluster name.Default: dev. Example: cluster.name=hzc1
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+	- cloud.token: A string. Sets the Hazelcast Cloud token. Example: cloud.token=1234567abcde
+	- stats.period: Duration, which can be parsed by time.Parse.
```
[the remark this answers is not in the document]
Comment (yuce) [reply sql/driver/doc.go]: Added at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go (hunk 1 above)
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go (hunk 2 above)
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go (hunk 4 above)
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go (hunk 5 above)
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go (hunk 6 above)
Comment (yuce) [reply sql/driver/doc.go:177]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go (hunk 7 above)
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/b0798c9d652c2a57683c730b872d5d2fc19d71c0

File: sql/driver/doc.go (hunk 2 above)
Comment (srknzl) [reply sql/driver/doc.go]: By adding a "reference" I did not mean a link. We can say that it is explained in the `Passing Hazelcast Specific Parameters` section below. 

Like this: 

They can be used to pass Hazelcast specific parameters, such as the cursor buffer size. **For details of how this works, check out Passing Hazelcast Specific Parameters section below.**

File: sql/driver/example_test.go (hunk 3 above)
Comment (srknzl) [reply sql/driver/example_test.go:27]: I did not see it, where should it appear? I use godoc and went to http://localhost:6060/pkg/github.com/hazelcast/hazelcast-go-client/sql/driver/

File: sql/driver/doc.go (hunk 2 above)
Comment (yuce) [reply sql/driver/doc.go]: OK, updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/fc11615214789dfe475c36ea8deff6d4400e1ccc

**Review by srknzl — APPROVED** [review]

File: sql/driver/doc.go [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+/*
+Package driver provides a standard database/sql compatible SQL driver for Hazelcast.
+
+This driver supports Hazelcast 5.0 and up. Check out the Hazelcast SQL documentation here: https://docs.hazelcast.com/hazelcast/latest/sql/sql-overview
+
+The documentation for the database/sql package is here: https://pkg.go.dev/database/sql
+
+Enabling Hazelcast SQL
+
+The SQL support should be enabled in Hazelcast configuration:
```
Comment (utku-caglayan) [inline sql/driver/doc.go:26]:
```suggestion
The SQL support should be enabled in Hazelcast member configuration:
```
I feel these configurations gets mixed up quite often from experiences on CLC. Maybe we can clarify that not the client config?

File: sql/driver/doc.go [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+	hz://dave:s3cr3t@my-server1.company.com:5000,my-server2.company.com:6000?cluster.name=prod&ssl=true&log=warn
+
+The following are the available options:
+
+	- unisocket: A boolean. Enables/disables the unisocket mode. Default: false. Example: unisocket=true
+	- log: One of the following: off, fatal, error, warn, info, debug, trace. Default: info. Example: log=debug
+	- cluster.name: A string. Specifies the cluster name. Default: dev. Example: cluster.name=hzc1
+	- cloud.token: A string. Sets the Hazelcast Cloud token. Example: cloud.token=1234567abcde
+	- stats.period: Duration between sending statistics, which can be parsed by time.Parse.
+	  Use one of the following suffixes: s (seconds), m (minutes), h (hours). Example: 10s
```
Comment (utku-caglayan) [inline sql/driver/doc.go:63]:
```suggestion
	  Use one of the following suffixes: s (seconds), m (minutes), h (hours). Example: stats.period=10s
```

**Review by utku-caglayan — APPROVED** [review]
Minor suggestions. Looks great!

File: sql/driver/doc.go (hunk 13 above)
Comment (yuce) [reply sql/driver/doc.go]: I agree, updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/f9a3272a3f7266e575f052f13da0ae342c3d20d8

File: sql/driver/doc.go (hunk 14 above)
Comment (yuce) [reply sql/driver/doc.go]: Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/709/commits/f9a3272a3f7266e575f052f13da0ae342c3d20d8

**Comment by yuce** [comment]
@srknzl @utku-caglayan Thanks for the review!

**Pull request merged**